pkg/scheduler/framework/parallelize/parallelism.go

```diff
@@ -51,14 +52,25 @@ func chunkSizeFor(n, parallelism int) int {
 	return s
 }
 
+type DoWorkPieceFuncWithError func(piece int) error
+
 // Until is a wrapper around workqueue.ParallelizeUntil to use in scheduling algorithms.
 // A given operation will be a label that is recorded in the goroutine metric.
-func (p Parallelizer) Until(ctx context.Context, pieces int, doWorkPiece workqueue.DoWorkPieceFunc, operation string) {
+func (p Parallelizer) Until(ctx context.Context, pieces int, doWorkPiece DoWorkPieceFuncWithError, operation string) {
```

Comment:
Looks like  `errCh.SendErrorWithCancel(someError, cancel)` is a current common pattern to handle errors in `doWorkPiece`. Can we follow it?

Reply:
@sanposhiho 
I see, but I'm not sure if there's a clean implementation for this
because we can't receive an error from errCh in two places at the same time.
One idea could be to define errCh within withMetrics, and when it receives an error, it could also forward that error to an external channel. What do you think about this approach?

```golang
func (p Parallelizer) Until(ctx context.Context, pieces int, doWorkPiece DoWorkPieceFuncWithError, operation string, errCh errorChannel) {
	goroutinesMetric := metrics.Goroutines.WithLabelValues(operation)
	withMetrics := func(piece int) {
		goroutinesMetric.Inc()
                anotherCh := errorChannel()
		doWorkPiece(piece, anotherCh)
                if err := anotherCh.ReceiveError(); err != nil {
                  errCh.SendErrorWithCancel(err)
```

Reply:
can't we just remove all our `errCh` usages from `doWorkPiece`s, and let them just return errors?

e.g., in framework.go:

```diff
-		f.Parallelizer().Until(ctx, len(nodes), func(index int) {
+		err := f.Parallelizer().Until(ctx, len(nodes), func(index int) {
...
				s, status := f.runScorePlugin(ctx, pl, state, pod, nodeName)
				if !status.IsSuccess() {
					err := fmt.Errorf("plugin %q failed with: %w", pl.Name(), status.AsError())
-					errCh.SendErrorWithCancel(err, cancel)
-                                       return
+					return err 
				}
...
-		if err := errCh.ReceiveError(); err != nil {
-			return nil, framework.AsStatus(fmt.Errorf("running Score plugins: %w", err))
-		}
```

```go
func (p Parallelizer) Until(ctx context.Context, pieces int, doWorkPiece DoWorkPieceFuncWithError, operation string) error {

	errCh := parallelize.NewErrorChannel()
	withMetrics := func(piece int) {
		if err := doWorkPiece(piece); err != nil {
			result = "error"
			errCh.SendErrorWithCancel(err, cancel)  // send errors to the ch.
		}
		// ...
	}
	workqueue.ParallelizeUntil(ctx, p.parallelism, pieces, withMetrics, workqueue.WithChunkSize(chunkSizeFor(pieces, p.parallelism)))

	if err := errCh.ReceiveError(); err != nil { // catch an error and return.
		return err
	}

	return nil
}
```

Reply:
@sanposhiho ,
OK, I've fixed it, could you review it again?